capirca/lib/fortigate.py

```diff
@@ -0,0 +1,937 @@
+      return f_proto
+    if port:
+      try:
+        return f_proto[port]
+      except KeyError as exc:
+        raise FortiGatePortDoesNotExistError(f"Port {exc} does not exist") from exc
+    else:
+      raise FortiGateFindServiceError(
+          f'service not found from {protocol} protocol and {port} port')
+
+
+class ObjectsContainer():
+  """A Container that holds service and network objects."""
+
+  def __init__(self):
+    self.verbose = True
+    self._sys_settings = set()
+    self._dict_addresses = set()
+    self._dict_addrgrps = {}
+    self._dict_services = {}
+    self._dict_svcgrps = {}
+    self._dict_schedules = {}
+
+  def get_sys_settings(self):
+    """Returns the collected addresses.
+
+    """
+    settings = []
+    for setting in self._sys_settings:
+      settings += [_SP + setting]
+
+    return settings
+
+  def get_fw_addresses(self, ip_v=4):
+    """
+    Returns the collected firewall addresses.
+
+    Args:
+        ip_v: an integer. version 4 or 6.
+
+    Returns: a list. contains address objects.
+
+    """
+    addresses = []
+    addresses_v = []
+    for addr in self._dict_addresses:
+      if ip_v == 4 and not isinstance(addr, nacaddr.IPv6):
+        addresses_v += [addr]
+      if ip_v == 6 and isinstance(addr, nacaddr.IPv6):
+        addresses_v += [addr]
+
+    for addr in sorted(addresses_v):
+      addresses.extend(self.get_fw_address_obj(addr, self.verbose))
+
+    return addresses
+
+  def get_fw_address_obj(self, addr, verbose=True):
+    """Gets firewall address objects.
+
+    """
+    addr_name = addr.with_prefixlen
+    address = []
+    address += [f'{_SP} edit "{addr_name}"']
+    if verbose:
+      addr_comment = _DEFAULT_COMMENT
+      if len(addr.text):
+        addr_comment += " (" + addr.text + ")"
+      address += [f'{_SP * 2} set comment "{self.fix_comment_length(addr_comment)}"']
+    if not isinstance(addr, nacaddr.IPv6):
+      address += [f'{_SP * 2} set subnet {addr_name}']
+    if isinstance(addr, nacaddr.IPv6):
+      address += [f'{_SP * 2} set ip6 {addr_name}']
+    address += [_SP + 'next']
+
+    return address
+
+  def get_fw_addrgrps(self, ip_v=4):
+    """Returns the collected address groups."""
+    addrgrps = []
+    for addrgrp_name, addrgrp_item in self._dict_addrgrps.items():
+      if addrgrp_item[0] == ip_v:
+        address = addrgrp_item[1]
+        exclude_address = addrgrp_item[2]
+        addrgrps += [f'{_SP} edit "{addrgrp_name}"']
+        if self.verbose:
+          addrgrps += [f'{_SP * 2} set comment "{_DEFAULT_COMMENT}"']
+        if address:
+          addrgrps += [f'{_SP * 2} set member {" ".join(f"\"{v}\"" for v in address)}']
+        else:
+          addrgrps += [_SP * 2 + 'set member "all"']
+        if exclude_address:
+          addrgrps += [_SP * 2 + 'set exclude enable']
+          addrgrps += [f'{_SP * 2} set exclude-member {" ".join(f"\"{v}\"" for v in exclude_address)}']
+        addrgrps += [_SP + 'next']
+
+    return addrgrps
```

Comment:
Is there a limit to the number of address groups supported by the platform? And should we error out if this is exceeded?

https://github.com/google/capirca/blob/master/doc/generator_patterns.md#check-various-limits-when-rendering-output